### PR TITLE
extract_all_objects(): Recursively extract objects

### LINKED ISSUE #3401: transitive extract_all_objects no longer works
There is new error added by commit 809f01833336be63eb07441e120de2e4f8c4b3c4: "extract_all_objects: Also extract generated sources". 

 ERROR:  Cannot extract objects from a target that itself has extracted objects

There are many cases where this might be needed, and it seemed to work fine in previous versions of meson. This is breaking the build of the DPDK project where we are compiling for older x86_64 targets without AVX2 instruction set support. In that case, we compile the AVX2 code separately as a static lib, extact it and then merge with the rest of the code into the official library static lib. That lib in turn is converted to a shared lib again by extract_all_objects(). [Yes, I have a TODO to look at SIMD module and "both_library" support for this, but it is what it is now.]

Please fix before next release. This used to work and is a regression IMHO.

### PR DESCRIPTION
**Pull request by xclaesse**
Fixes #3401

### CODE REVIEW

**Comment by bruce-richardson** [comment]
I've tried a quick test with this, and while it could be my project meson.build files screwing up here, it's failing the link stage. From what I can see it appears that the .o file from the recursive object extraction is getting passed twice on the commandline causing duplicate symbol errors.

**Comment by bruce-richardson** [comment]
Tested and works ok for me.
However, this is a behaviour change, and I think needs to be flagged in the release notes. I have to put a meson version check into my files to take account of this.

**Comment by bruce-richardson** [comment]
FYI: here was my attempt at a fix: https://github.com/bruce-richardson/meson/commit/a6d19edb4b575c764bfda2bc1ebe0d74cb38dac7

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -398,36 +396,43 @@ def object_filename_from_source(self, target, source, is_unity):
...
+        sources = list(extobj.srclist)
+        for gensrc in extobj.genlist:
+            for s in gensrc.get_outputs():
+                path = self.get_target_generated_dir(extobj.target, gensrc, s)
+                dirpart, fnamepart = os.path.split(path)
+                sources.append(File(True, dirpart, fnamepart))
+
+        # Filter out headers and all non-source files
+        sources = [s for s in sources if self.environment.is_source(rel_src) and \
+                                         not self.environment.is_header(rel_src)]
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

[E127] continuation line over-indented for visual indent

File: mesonbuild/build.py [hunk 2]
```diff
@@ -228,14 +229,16 @@ def check_unity_compatible(self):
...
         # from each unified source file.
-        # If the list of sources for which we want objects is the same as the
-        # list of sources that go into each unified build, we're good.
+        #
+        # FIXME: Due to limitations in backends.determine_ext_objs() all
+        # extracted sources must have the same compiler. We also should check
+        # here which compiler is needed for each generated source, forbid
+        # generated source for now.
+        if self.genlist:
+            raise MesonException('Cannot extract objects from Unity builds '\
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[Flake8]__

[E502] the backslash is redundant between brackets

File: mesonbuild/backend/backends.py (hunk 1 above)
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

[F821] undefined name 'rel_src'

File: mesonbuild/backend/backends.py [hunk 3]
```diff
@@ -398,36 +396,43 @@ def object_filename_from_source(self, target, source, is_unity):
...
+        # Merge sources and generated sources
+        sources = list(extobj.srclist)
+        for gensrc in extobj.genlist:
+            for s in gensrc.get_outputs():
+                path = self.get_target_generated_dir(extobj.target, gensrc, s)
+                dirpart, fnamepart = os.path.split(path)
+                sources.append(File(True, dirpart, fnamepart))
+
+        # Filter out headers and all non-source files
+        sources = [s for s in sources if self.environment.is_source(rel_src) and \
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

[E502] the backslash is redundant between brackets

Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

[F821] undefined name 'rel_src'

File: mesonbuild/backend/backends.py [hunk 4]
```diff
@@ -398,36 +402,43 @@ def object_filename_from_source(self, target, source, is_unity):
...
+        sources = list(extobj.srclist)
+        for gensrc in extobj.genlist:
+            for s in gensrc.get_outputs():
+                path = self.get_target_generated_dir(extobj.target, gensrc, s)
+                dirpart, fnamepart = os.path.split(path)
+                sources.append(File(True, dirpart, fnamepart))
+
+        # Filter out headers and all non-source files
+        sources = [s for s in sources if self.environment.is_source(s) and \
+                                         not self.environment.is_header(s)]
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

[E127] continuation line over-indented for visual indent

File: mesonbuild/backend/backends.py [hunk 5]
```diff
@@ -398,36 +402,43 @@ def object_filename_from_source(self, target, source, is_unity):
...
+        # Merge sources and generated sources
+        sources = list(extobj.srclist)
+        for gensrc in extobj.genlist:
+            for s in gensrc.get_outputs():
+                path = self.get_target_generated_dir(extobj.target, gensrc, s)
+                dirpart, fnamepart = os.path.split(path)
+                sources.append(File(True, dirpart, fnamepart))
+
+        # Filter out headers and all non-source files
+        sources = [s for s in sources if self.environment.is_source(s) and \
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

[E502] the backslash is redundant between brackets

**Comment by bruce-richardson** [comment]
Question: do you have any idea how many projects may be expecting the old behaviour which was to silently drop the recursive dependencies. The DPDK project currently expects that and now requires a version check in the code.

If it's likely that a number of other projects are expecting this current behaviour, can I suggest adding an additional parameter to extract_all_objects to indicate whether or not to return recursive objects. That would keep compatibility with previous versions (assuming it defaults to not returning).

**Comment by xclaesse** [comment]
I had no idea, why would you not want recursive objects? That seems bogus to me.

**Comment by bruce-richardson** [comment]
Oh, we would want recursive objects and the new behaviour is correct but unfortunately it is different. Changing the behaviour will break code build around the old one.

Not only that, it’s not possible to fix it in a way that works correctly with meson 0.45 and 0.46 so one has no choice but to put in an explicit version check. For example: http://dpdk.org/browse/dpdk/tree/lib/meson.build#n98 works fine now, but for v0.46 onwards the += needs to be =. I can’t see any way to avoid using hardcoded version checks.

It’s probably not a big deal but it is an incompatible change.


On 13 Apr 2018, at 23:16, Xavier Claessens <notifications@github.com<mailto:notifications@github.com>> wrote:


I had no idea, why would you not want recursive objects? That seems bogus to me.

**Comment by xclaesse** [comment]
Why don't you use extract_objects() if you don't want all objects ?

**Comment by xclaesse** [comment]
Oh I understand, you had a workaround the missing objects, and now that the root cause is fixed you need to drop your hack... I guess it's up to @jpakkan to decide, personally I think we should fix it in meson because that's how it always should have worked. If you can't hard dep on meson >=0.46 one releases, it's not that hard to add a version check, is it?

**Comment by bruce-richardson** [comment]
Yes, it is the case that we had workarounds (though in practice it just happened to work first try with the code i had) which are now broken. 

I'm fine to have this merged, and it's no big deal for me to put in a version check, I just wanted to flag that this may break other projects code too.

Final point - if this doesn't get merged, we still need some change in this area, as the current latest code completely errors out if you call extract_all_objects() with recursive objects in it. The old behaviour at least was usable with workarounds, the error case isn't.

**Comment by jpakkane** [comment]
What about adding a `recursive` keyword to define whether recursive dependencies should be added or not? And have it default to the old behaviour?

**Comment by xclaesse** [comment]
That would work, yes. But it means we'll have to keep a broken behaviour by default forever. I guess it's a bit mitigated by the fact that with both_libraries() we much less need to use extract_all_objects().

**Comment by jpakkane** [comment]
But on the other hand, is the non-recursive extraction a thing that some projects would need on its own? If yes, it would make sense to provide this feature in general. If not, then we could consider making the new one the default, having `false` deprecated and eventually remove the kwarg altogether.

**Comment by xclaesse** [comment]
I don't think it makes sense to be non-recursive, the project @bruce-richardson is working on needs recursive but because it used to not be recursive he added manually objects. And if we start being recursive it means those objects become duplicated and it fails to build.

So if we think bug-for-bug backward compat is important, then we need to add a `recursive` keyword arg that default to `false` and everyone will have to add `recursive: true`.

**Comment by bruce-richardson** [comment]
+1 for adding option. However, I'd also point out that an option allows a planned transition to having recursive by default. You can check and omit a warning for projects that haven't specified a value. Then after a few releases, once projects have had time to update, you can change the default.

**Comment by jpakkane** [comment]
I can't see a way to maintain backwards compatibility without `recursive` so let's add that in.

File: mesonbuild/interpreter.py [hunk 6]
```diff
@@ -587,8 +587,15 @@ def extract_objects_method(self, args, kwargs):
         gobjs = self.held_object.extract_objects(args)
         return GeneratedObjectsHolder(gobjs)
 
+    @permittedMethodKwargs({'recursive'})
     def extract_all_objects_method(self, args, kwargs):
-        gobjs = self.held_object.extract_all_objects()
+        recursive = kwargs.get('recursive', False)
+        gobjs = self.held_object.extract_all_objects(recursive)
+        if gobjs.objlist and not 'recursive' in kwargs:
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__

[E713] test for membership should be 'not in'

**Pull request merged**